snuba/query/query.py

```diff
@@ -204,3 +219,138 @@ def get_all_referenced_columns(self) -> Sequence[Any]:
+        returned by that function and it will replace all the columns reteturned by that
+        function.
+        """
+        if isinstance(expression, str):
+            match = ORDERED_COL.match(expression)
+            if match and match[1] == old_column:
+                return expression.replace(old_column, new_column)
+        elif (isinstance(expression, (list, tuple)) and len(expression) >= 2
+                and isinstance(expression[1], (list, tuple))):
+            params = [
+                self.__replace_col_in_expression(param, old_column, new_column)
+                for param in expression[1]
+            ]
+            ret = [expression[0], params]
+            if len(expression) == 3:
+                # Alias for the column
+                ret.append(expression[2])
+            return ret
+        return expression
+
+    def __replace_col_in_condition(
+        self,
+        condition: Condition,
+        old_column: str,
+        new_column: str,
+    ) -> Condition:
+        """
+        Replaces a column in a structured condition. This is a level above replace_col_in_expression
+        since conditions are in the form [expression, operator, literal] (which is not fully correct
+        since the right side of the condition should be an expression as well but this constraint is
+        imposed by the query schema and get_all_referenced_columns behaves accordingly).
+        Conditions can also be nested.
+        """
+        if is_condition(condition):
+            return [
+                self.__replace_col_in_expression(condition[0], old_column, new_column),
+                condition[1],
+                condition[2]
+            ]
+        elif isinstance(condition, (tuple, list)):
+            # nested condition
+            return [
+                self.__replace_col_in_condition(cond, old_column, new_column) for cond in condition
+            ]
+        else:
+            # Don't know what this is
+            return condition
+
+    def __replace_col_in_list(
+        self,
+        expressions: Any,
+        old_column: str,
+        new_column: str,
+    ) -> Sequence[Any]:
+        return [
+            self.__replace_col_in_expression(expr, old_column, new_column)
+            for expr in to_list(expressions)
+        ]
+
+    def replace_column(self, old_column: str, new_column: str) -> None:
+        """
+        Replaces a column in all fields of the query. The Query object is mutated in place
+        while the internal fields are replaced.
+
+        This behaves consistently with get_all_referenced_columns (which does not really
+        behave correctly since it is missing a few fields that can contain columns). Will
+        fix both when adding a better column abstraction.
+
+        In the current implementation we can only replace a column identified by a string
+        with another column identified by a string. This does not support replacing a
+        column with a complex expression.
```

Comment:
It actually can replace a tag with something else since a tag is a "column represented as a string". Will add a test about that.

Reply:
It's not obvious to me how that'd work because of the regular expression — curious to see if I'm missing something or if it will require changes to support that.

Reply:
I will clarify in the comment. 
This is really not that sophisticated at this level for two reasons:
- this method is always going to be called before any Clickhouse formatting 
- column_expr (which resolves tags) is performed during Clickhosue formatting right now. It will change soon. 

This implies that this method will encounter tags in the form of "tag[aaaa]" which is a string. Potentially this method can be used to replace it with "not_a_tag[not_aaaa]" though it would not be a great idea so it will be discouraged but there is no way we can strictly prevent that without any structure around a tag column. column_expr will be executed later and pick up whatever it sees in the query.

With structured columns things will be better. There are two possible ways to deal with that, I am still thinking over which one is better:
1) resolve tags through a query processor early on.  This turns tags into standard columns (promoted) or into function calls. This is straightforward. But it makes the concept of tag disappear for the following processors. They would not know that the function call came from a tag resolution or to replace a tag with a different tag.
2) parse tag[a] into a specific column type we can manipulate and resolve it right before Clickhouse formatting. This is more complex, it requires dataset to be able to enable/disable some feature during parsing, but it allows query processors to manipulate a higher level concept (the tag) thus being more powerful.
Will resolve this problem in the next days.